java-frontend/src/main/java/org/sonar/java/model/JParser.java

```diff
@@ -2082,8 +2081,18 @@ private ExpressionTree createExpression(Expression node) {
         InstanceofExpression e = (InstanceofExpression) node;
         return new InstanceOfTreeImpl(
           firstTokenAfter(e.getLeftOperand(), TerminalTokens.TokenNameinstanceof),
-          convertType(e.getRightOperand()),
-          e.getAST().isPreviewEnabled() && e.getPatternVariable() != null ? convertSimpleName(e.getPatternVariable()) : null
+          convertType(e.getRightOperand())
+        ).complete(
+          convertExpression(e.getLeftOperand())
+        );
+      }
+      case ASTNode.PATTERN_INSTANCEOF_EXPRESSION: {
+        PatternInstanceofExpression e = (PatternInstanceofExpression) node;
+        SingleVariableDeclaration v = e.getRightOperand();
+        return new InstanceOfTreeImpl(
+          firstTokenAfter(e.getLeftOperand(), TerminalTokens.TokenNameinstanceof),
+          convertType(v.getType()),
+          convertSimpleName(v.getName())
```

Comment:
We should also ensure that `InstanceOfTree#patternVariable().symbol().declaration()` matches this `InstanceOfTree`

Reply:
I introduced a new KIND (`PATTERN_INSTANCE_OF`) and Tree interface (`PatternInstanceOfTree`) to do that.